# Student grades prediction

In [2]:
# import necessary libraries
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [3]:
#load dataset
df = pd.read_csv('Grades.csv')

In [4]:
df.head()

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102 EE-119  \
0  CS-97001     B-     D+     C-         C     C-     D+      D     C-     B-   
1  CS-97002      A      D     D+         D     B-      C      D      A     D+   
2  CS-97003      A      B      A        B-     B+      A     B-     B+     A-   
3  CS-97004      D     C+     D+         D      D     A-     D+     C-      D   
4  CS-97005     A-     A-     A-        B+      A      A     A-     B+      A   

   ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423 CS-412   CGPA  
0  ...     C-     C-     C-     C-     A-      A     C-      B     A-  2.205  
1  ...     D+      D      C      D     A-     B-      C      C      B  2.008  
2  ...      B      B      A      C      A      A      A     A-      A  3.608  
3  ...     D+      C     D+     C-     B-      B     C+     C+     C+  1.906  
4  ...     B-     B+     B+     B-     A-      A     A-     A-      A  3.448  

[5 rows x 43 columns]

In [5]:
df.shape

(571, 43)

Here we observed that 571 rows and 43 columns are there in out dataset.

In [7]:
df.columns

Index(['Seat No.', 'PH-121', 'HS-101', 'CY-105', 'HS-105/12', 'MT-111',
       'CS-105', 'CS-106', 'EL-102', 'EE-119', 'ME-107', 'CS-107', 'HS-205/20',
       'MT-222', 'EE-222', 'MT-224', 'CS-210', 'CS-211', 'CS-203', 'CS-214',
       'EE-217', 'CS-212', 'CS-215', 'MT-331', 'EF-303', 'HS-304', 'CS-301',
       'CS-302', 'TC-383', 'MT-442', 'EL-332', 'CS-318', 'CS-306', 'CS-312',
       'CS-317', 'CS-403', 'CS-421', 'CS-406', 'CS-414', 'CS-419', 'CS-423',
       'CS-412', 'CGPA'],
      dtype='object')

This are the columns names present in our dataset.

In [8]:
#information of dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 571 entries, 0 to 570
Data columns (total 43 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Seat No.   571 non-null    object 
 1   PH-121     571 non-null    object 
 2   HS-101     571 non-null    object 
 3   CY-105     570 non-null    object 
 4   HS-105/12  570 non-null    object 
 5   MT-111     569 non-null    object 
 6   CS-105     571 non-null    object 
 7   CS-106     569 non-null    object 
 8   EL-102     569 non-null    object 
 9   EE-119     569 non-null    object 
 10  ME-107     569 non-null    object 
 11  CS-107     569 non-null    object 
 12  HS-205/20  566 non-null    object 
 13  MT-222     566 non-null    object 
 14  EE-222     564 non-null    object 
 15  MT-224     564 non-null    object 
 16  CS-210     564 non-null    object 
 17  CS-211     566 non-null    object 
 18  CS-203     566 non-null    object 
 19  CS-214     565 non-null    object 
 20  EE-217    

Here we can see that the above dataset having more number of object datatype and with object datatype we can't build the model so we need to convert all the column into numerical column. 

In [9]:
#check null values
df.isnull().sum()

Seat No.      0
PH-121        0
HS-101        0
CY-105        1
HS-105/12     1
MT-111        2
CS-105        0
CS-106        2
EL-102        2
EE-119        2
ME-107        2
CS-107        2
HS-205/20     5
MT-222        5
EE-222        7
MT-224        7
CS-210        7
CS-211        5
CS-203        5
CS-214        6
EE-217        6
CS-212        6
CS-215        6
MT-331        9
EF-303       10
HS-304       10
CS-301       10
CS-302       10
TC-383       10
MT-442       10
EL-332        9
CS-318        9
CS-306        9
CS-312       10
CS-317       12
CS-403       12
CS-421       12
CS-406       85
CS-414       13
CS-419       13
CS-423       14
CS-412       79
CGPA          0
dtype: int64

As we can see there are many null values in our dataset first we need to handel that null values.

In [11]:
# Dropping rows with any missing values
df = df.dropna()

# Dropping columns with more than 50% missing values
threshold = len(df) * 0.5
df = df.dropna(thresh=threshold, axis=1)

In [12]:
df.shape

(417, 43)

In [13]:
df.isnull().sum()

Seat No.     0
PH-121       0
HS-101       0
CY-105       0
HS-105/12    0
MT-111       0
CS-105       0
CS-106       0
EL-102       0
EE-119       0
ME-107       0
CS-107       0
HS-205/20    0
MT-222       0
EE-222       0
MT-224       0
CS-210       0
CS-211       0
CS-203       0
CS-214       0
EE-217       0
CS-212       0
CS-215       0
MT-331       0
EF-303       0
HS-304       0
CS-301       0
CS-302       0
TC-383       0
MT-442       0
EL-332       0
CS-318       0
CS-306       0
CS-312       0
CS-317       0
CS-403       0
CS-421       0
CS-406       0
CS-414       0
CS-419       0
CS-423       0
CS-412       0
CGPA         0
dtype: int64

We droped the null all null value from the data which are having more null values.

In [19]:
# checking the value count for each column
for i in df.columns:
    print(df[i].value_counts())
    print('\n')

Seat No.
CS-97001    1
CS-97217    1
CS-97437    1
CS-97436    1
CS-97435    1
           ..
CS-97142    1
CS-97141    1
CS-97140    1
CS-97139    1
CS-97571    1
Name: count, Length: 417, dtype: int64


PH-121
A     105
A-    103
B+     50
B-     43
B      36
A+     22
C      17
C+     14
C-     11
D+      9
D       7
Name: count, dtype: int64


HS-101
A-    64
B-    62
B     47
C     46
C-    40
B+    40
C+    36
D     30
A     26
D+    23
A+     3
Name: count, dtype: int64


CY-105
A     152
A-    108
B+     39
B-     26
B      25
A+     16
C      13
C-     11
C+     10
D+      9
D       8
Name: count, dtype: int64


HS-105/12
A     81
A-    62
B+    42
B     37
A+    33
B-    32
D     28
D+    27
C     26
C+    26
C-    23
Name: count, dtype: int64


MT-111
A-    82
A     59
B-    56
B     45
B+    44
C-    29
C+    25
D     23
C     20
D+    18
A+    16
Name: count, dtype: int64


CS-105
A     135
A-    120
B+     45
A+     41
B      25
B-     16
C+     12
C       7
D+      6
C-  

Here in each column there are several categories in object datatype so we need to convert those datatype in number data for model building.

we will assign some numbers to to the several category in our dataset.

In [20]:
# Define the values that need to be dropped
values_to_drop = ['W', 'WU']

# Drop rows containing 'W' or 'WU' in any column
df = df[~df.isin(values_to_drop).any(axis=1)]

# Display the first few rows of the cleaned dataset
df.shape

(408, 43)

Grade A+:- 4.3, A:- 4.0, A-:- 3.7, B+:- 3.3, B:-3.0, B-:-2.7, C+:- 2.3, C:- 2.0, C-:- 1.7, D+:- 1.3, D:- 1.0, D-:- 0.7, F:-0.0

In [21]:
# Define a mapping for letter grades to numerical values
grade_mapping = {
    'A+': 4.3, 'A': 4.0, 'A-': 3.7, 'B+': 3.3, 'B': 3.0, 'B-': 2.7,
    'C+': 2.3, 'C': 2.0, 'C-': 1.7, 'D+': 1.3, 'D': 1.0, 'D-': 0.7, 'F': 0.0,
    'A+': 4.0, 'C+R': 2.3 }

# Apply the mapping to the relevant columns
for col in df.columns[1:-1]:  # Exclude 'Seat No.' and 'CGPA'
    df[col] = df[col].map(grade_mapping)

In [22]:
df.head()

Seat No.  PH-121  HS-101  CY-105  HS-105/12  MT-111  CS-105  CS-106  \
0  CS-97001     2.7     1.3     1.7        2.0     1.7     1.3     1.0   
1  CS-97002     4.0     1.0     1.3        1.0     2.7     2.0     1.0   
2  CS-97003     4.0     3.0     4.0        2.7     3.3     4.0     2.7   
3  CS-97004     1.0     2.3     1.3        1.0     1.0     3.7     1.3   
4  CS-97005     3.7     3.7     3.7        3.3     4.0     4.0     3.7   

   EL-102  EE-119  ...  CS-312  CS-317  CS-403  CS-421  CS-406  CS-414  \
0     1.7     2.7  ...     1.7     1.7     1.7     1.7     3.7     4.0   
1     4.0     1.3  ...     1.3     1.0     2.0     1.0     3.7     2.7   
2     3.3     3.7  ...     3.0     3.0     4.0     2.0     4.0     4.0   
3     1.7     1.0  ...     1.3     2.0     1.3     1.7     2.7     3.0   
4     3.3     4.0  ...     2.7     3.3     3.3     2.7     3.7     4.0   

   CS-419  CS-423  CS-412   CGPA  
0     1.7     3.0     3.7  2.205  
1     2.0     2.0     3.0  2.008  
2     4.0     3.7     4.0  3.608  
3     2.3     2.3     2.3  1.906  
4     3.7     3.7     4.0  3.448  

[5 rows x 43 columns]

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 408 entries, 0 to 570
Data columns (total 43 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Seat No.   408 non-null    object 
 1   PH-121     408 non-null    float64
 2   HS-101     408 non-null    float64
 3   CY-105     408 non-null    float64
 4   HS-105/12  408 non-null    float64
 5   MT-111     408 non-null    float64
 6   CS-105     408 non-null    float64
 7   CS-106     408 non-null    float64
 8   EL-102     408 non-null    float64
 9   EE-119     408 non-null    float64
 10  ME-107     408 non-null    float64
 11  CS-107     408 non-null    float64
 12  HS-205/20  408 non-null    float64
 13  MT-222     408 non-null    float64
 14  EE-222     408 non-null    float64
 15  MT-224     408 non-null    float64
 16  CS-210     408 non-null    float64
 17  CS-211     408 non-null    float64
 18  CS-203     408 non-null    float64
 19  CS-214     408 non-null    float64
 20  EE-217     408 

Perfect our data set is now perfect now we can proceed ahead.

In [24]:
#Let's check statastic of the dataframe
df.describe()

PH-121      HS-101      CY-105   HS-105/12      MT-111      CS-105  \
count  408.000000  408.000000  408.000000  408.000000  408.000000  408.000000   
mean     3.328676    2.644363    3.437500    2.965686    2.954902    3.553922   
std      0.757265    0.882128    0.766363    1.003727    0.917038    0.640166   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      3.000000    2.000000    3.300000    2.300000    2.300000    3.300000   
50%      3.700000    2.700000    3.700000    3.300000    3.000000    3.700000   
75%      4.000000    3.300000    4.000000    4.000000    3.700000    4.000000   
max      4.000000    4.000000    4.000000    4.000000    4.000000    4.000000   

           CS-106     EL-102      EE-119      ME-107  ...      CS-312  \
count  408.000000  408.00000  408.000000  408.000000  ...  408.000000   
mean     2.882108    2.93799    3.179167    2.825000  ...    2.656863   
std      0.909707    0.91898    0.719188    0.991543  ...    0.972371   
min      1.000000    1.00000    1.000000    1.000000  ...    0.000000   
25%      2.300000    2.30000    2.700000    2.000000  ...    1.700000   
50%      3.000000    3.00000    3.300000    3.000000  ...    2.700000   
75%      3.700000    3.70000    3.700000    3.700000  ...    3.700000   
max      4.000000    4.00000    4.000000    4.000000  ...    4.000000   

           CS-317      CS-403      CS-421      CS-406      CS-414      CS-419  \
count  408.000000  408.000000  408.000000  408.000000  408.000000  408.000000   
mean     2.723284    3.111275    2.433578    3.350245    3.411275    3.074265   
std      0.856266    0.865966    0.870129    0.753403    0.671878    0.733407   
min      0.000000    1.000000    0.000000    0.000000    0.000000    0.000000   
25%      2.000000    2.700000    1.700000    3.000000    3.000000    2.700000   
50%      2.700000    3.300000    2.300000    3.700000    3.700000    3.300000   
75%      3.300000    4.000000    3.000000    3.700000    4.000000    3.700000   
max      4.000000    4.000000    4.000000    4.000000    4.000000    4.000000   

           CS-423      CS-412        CGPA  
count  408.000000  408.000000  408.000000  
mean     2.754657    3.289461    3.006574  
std      0.865147    0.734387    0.571054  
min      0.000000    0.000000    1.708000  
25%      2.000000    3.000000    2.607750  
50%      3.000000    3.700000    3.049500  
75%      3.700000    3.700000    3.470750  
max      4.000000    4.000000    3.985000  

[8 rows x 42 columns]

we will drop the Seat No. column we don't see and relationship with other column.

In [25]:
df = df.drop(columns = ['Seat No.'])

In [26]:
df.head()

PH-121  HS-101  CY-105  HS-105/12  MT-111  CS-105  CS-106  EL-102  EE-119  \
0     2.7     1.3     1.7        2.0     1.7     1.3     1.0     1.7     2.7   
1     4.0     1.0     1.3        1.0     2.7     2.0     1.0     4.0     1.3   
2     4.0     3.0     4.0        2.7     3.3     4.0     2.7     3.3     3.7   
3     1.0     2.3     1.3        1.0     1.0     3.7     1.3     1.7     1.0   
4     3.7     3.7     3.7        3.3     4.0     4.0     3.7     3.3     4.0   

   ME-107  ...  CS-312  CS-317  CS-403  CS-421  CS-406  CS-414  CS-419  \
0     1.7  ...     1.7     1.7     1.7     1.7     3.7     4.0     1.7   
1     1.0  ...     1.3     1.0     2.0     1.0     3.7     2.7     2.0   
2     3.7  ...     3.0     3.0     4.0     2.0     4.0     4.0     4.0   
3     2.3  ...     1.3     2.0     1.3     1.7     2.7     3.0     2.3   
4     3.7  ...     2.7     3.3     3.3     2.7     3.7     4.0     3.7   

   CS-423  CS-412   CGPA  
0     3.0     3.7  2.205  
1     2.0     3.0  2.008  
2     3.7     4.0  3.608  
3     2.3     2.3  1.906  
4     3.7     4.0  3.448  

[5 rows x 42 columns]

Here in our dataset we have many columns so we need to select top best features from 42 columns.

In [27]:
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.model_selection import train_test_split

In [28]:
# Here our label column is CGPA
X = df.drop('CGPA', axis=1)
y = df.CGPA

In [37]:
best_feature = SelectKBest(score_func = f_regression, k = 23)
fit = best_feature.fit(X,y)
df_scores = pd.DataFrame(fit.scores_)
df_columns = pd.DataFrame(X.columns)

In [38]:
# concatnate dataframe
feature_score = pd.concat([df_columns,df_scores],axis=1)

In [39]:
feature_score.columns = ['Feature_Name', 'Score'] #name output column
print(feature_score.nlargest(23,'Score')) #print 25 best features

   Feature_Name       Score
15       CS-210  953.451217
33       CS-317  857.070336
29       EL-332  790.547197
26       CS-302  765.529042
25       CS-301  705.713027
21       CS-215  670.275648
35       CS-421  611.596241
20       CS-212  592.623139
18       CS-214  581.930607
27       TC-383  568.142189
38       CS-419  561.704193
14       MT-224  546.389069
4        MT-111  532.014907
0        PH-121  506.468623
31       CS-306  487.250962
17       CS-203  435.625007
12       MT-222  428.483913
39       CS-423  400.403281
32       CS-312  378.580010
28       MT-442  367.158540
19       EE-217  342.705243
8        EE-119  336.704406
16       CS-211  328.373073


In [40]:
new_X = df[['CS-210', 'CS-317', 'EL-332', 'CS-302', 'CS-301', 'CS-215', 'CS-421', 'CS-212', 'CS-214', 'TC-383', 'CS-419', 'MT-224', 'MT-111', 'PH-121', 'CS-306', 'CS-203', 'MT-222', 'CS-423', 'CS-312', 'MT-442', 'EE-217', 'EE-119', 'CS-211']]

In [41]:
new_X

CS-210  CS-317  EL-332  CS-302  CS-301  CS-215  CS-421  CS-212  CS-214  \
0       2.3     1.7     2.0     2.0     3.0     1.7     1.7     1.0     1.0   
1       1.3     1.0     2.0     1.3     3.7     1.0     1.0     2.0     1.0   
2       4.0     3.0     3.3     3.7     3.3     4.0     2.0     3.7     3.7   
3       2.0     2.0     1.3     2.7     1.0     1.0     1.7     1.0     1.0   
4       3.7     3.3     3.7     3.7     2.7     3.7     2.7     3.7     3.7   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
566     4.0     3.7     4.0     4.0     3.7     3.7     4.0     3.0     4.0   
567     4.0     3.3     4.0     4.0     4.0     3.0     4.0     4.0     4.0   
568     3.7     3.0     3.0     3.3     3.3     3.0     3.3     3.3     3.0   
569     2.7     3.0     1.0     2.3     2.0     1.3     1.7     1.3     2.0   
570     3.3     2.0     0.0     1.0     1.0     1.3     1.0     2.0     1.3   

     TC-383  ...  PH-121  CS-306  CS-203  MT-222  CS-423  CS-312  MT-442  \
0       2.3  ...     2.7     2.0     1.3     1.0     3.0     1.7     3.3   
1       1.7  ...     4.0     1.0     2.0     2.7     2.0     1.3     1.7   
2       3.0  ...     4.0     3.7     4.0     4.0     3.7     3.0     4.0   
3       1.3  ...     1.0     1.7     2.3     1.3     2.3     1.3     1.7   
4       2.3  ...     3.7     3.7     4.0     4.0     3.7     2.7     3.7   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
566     4.0  ...     3.0     3.3     4.0     4.0     3.0     3.7     4.0   
567     4.0  ...     4.0     3.7     4.0     2.3     2.0     3.3     4.0   
568     4.0  ...     3.0     3.3     3.3     3.0     3.7     3.7     3.3   
569     1.7  ...     4.0     3.3     2.7     1.0     2.7     1.0     1.7   
570     2.0  ...     2.0     1.3     3.0     1.7     1.0     2.3     3.7   

     EE-217  EE-119  CS-211  
0       4.0     2.7     1.3  
1       2.7     1.3     1.0  
2       4.0     3.7     4.0  
3       1.3     1.0     3.3  
4       4.0     4.0     2.7  
..      ...     ...     ...  
566     4.0     4.0     4.0  
567     4.0     4.0     4.0  
568     4.0     4.0     4.0  
569     2.7     2.7     3.0  
570     1.7     2.0     1.7  

[408 rows x 23 columns]

In [43]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaler = scaler.fit_transform(new_X)

In [44]:
x_train, x_test, y_train, y_test = train_test_split(X_scaler, y, test_size=0.25, random_state=215)

# Model Building

In [45]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.ensemble import AdaBoostRegressor

In [47]:
print('Training feature matrix size:',x_train.shape)
print('Training target vector size:',y_train.shape)
print('Test feature matrix size:',x_test.shape)
print('Test target vector size:',y_test.shape)

Training feature matrix size: (306, 23)
Training target vector size: (306,)
Test feature matrix size: (102, 23)
Test target vector size: (102,)


In [48]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
maxR2_score=0
maxRS=0
for i in range(1,200):
    x_train,x_test,y_train,y_test = train_test_split(X_scaler,y,test_size = 0.25, random_state=i)
    lin_reg=LinearRegression()
    lin_reg.fit(x_train,y_train)
    y_pred=lin_reg.predict(x_test)
    R2=r2_score(y_test,y_pred)
    if R2>maxR2_score:
        maxR2_score=R2
        maxRS=i
print('Best accuracy is', maxR2_score ,'on Random_state', maxRS)

Best accuracy is 0.9892923171988318 on Random_state 144


# Linear Regression

In [50]:
x_train,x_test,y_train,y_test = train_test_split(X_scaler,y,test_size = 0.25, random_state=148)
lin_reg=LinearRegression()
lin_reg.fit(x_train,y_train)
lin_reg.score(x_train,y_train)
y_pred=lin_reg.predict(x_test)
print('Predicted result :\n', y_pred)
print('\n')
print('actual result\n',y_test)

Predicted result :
 [2.46364028 2.28085606 2.98590299 2.10654936 3.68992867 2.59467912
 2.65970618 3.14998721 2.9597949  3.77045325 3.56314816 2.47546756
 3.26732586 3.64567068 2.49836887 3.16502036 2.6135356  3.51046591
 3.25465447 3.7734246  1.81075904 2.97171306 3.72833221 2.67858061
 2.95069565 3.42980645 3.59122569 3.17586352 3.84172158 2.80918323
 2.8257216  2.34230424 3.08753656 2.72452434 3.13423933 2.4164919
 2.21978439 2.88866077 3.6868699  2.24017599 3.02638771 3.29604057
 2.61039739 3.83598138 2.4135518  2.73072519 2.4507311  1.81437443
 2.23090191 2.99019862 3.19381578 3.37562211 2.84098265 3.30813893
 3.92320479 3.69234344 3.35677293 2.5057833  3.03505684 3.5638298
 2.35562018 3.38479322 2.73653077 3.55273265 3.56057452 3.8456827
 2.69721666 2.10137631 3.02408905 2.81764549 2.58725424 3.74856395
 1.99733216 2.88919156 3.15203879 1.88470027 3.42675821 3.53198432
 3.67808118 3.23358382 3.01947308 2.58912863 2.04948091 3.40880426
 3.56627436 2.93592468 3.01482952 2.64777725 

# Evaluation Metrics of linear regression

In [51]:
print('\033[1m'+' Error :'+'\033[0m')
print('Mean absolute error :', mean_absolute_error(y_test,y_pred))
print('Mean squared error :', mean_squared_error(y_test,y_pred))
print('Root Mean Squared Error:', np.sqrt(mean_squared_error(y_test,y_pred)))
print('\n')
from sklearn.metrics import r2_score
print('\033[1m'+' R2 Score :'+'\033[0m')
print(r2_score(y_test,y_pred))

 Error :
Mean absolute error : 0.06339001122404299
Mean squared error : 0.0065250854708743715
Root Mean Squared Error: 0.08077800115671575


 R2 Score :
0.9790588782242744


In [52]:
rf = RandomForestRegressor(n_estimators = 100 )
svr=SVR()
dtc = DecisionTreeRegressor()
ad=AdaBoostRegressor()

model = [rf,svr,dtc,ad]

for m in model:
    m.fit(x_train,y_train)
    m.score(x_train,y_train)
    y_pred = m.predict(x_test)
                                            
    print('\033[1m'+' Error of ', m, ':' +'\033[0m')
    print('Mean absolute error :', mean_absolute_error(y_test,y_pred))
    print('Mean squared error :', mean_squared_error(y_test,y_pred))
    print('Root Mean Squared Error:', np.sqrt(mean_squared_error(y_test,y_pred)))
    print('\n')

    print('\033[1m'+' R2 Score :'+'\033[0m')
    print(r2_score(y_test,y_pred)) 
    print('\n')

 Error of  RandomForestRegressor() :
Mean absolute error : 0.08941401960784295
Mean squared error : 0.014024911059803892
Root Mean Squared Error: 0.1184268173168725


 R2 Score :
0.9549894983432734


 Error of  SVR() :
Mean absolute error : 0.07968495156276552
Mean squared error : 0.009857756465998286
Root Mean Squared Error: 0.09928623502781384


 R2 Score :
0.9683632529395428


 Error of  DecisionTreeRegressor() :
Mean absolute error : 0.18631372549019606
Mean squared error : 0.05712464705882353
Root Mean Squared Error: 0.23900762970839137


 R2 Score :
0.8166684259089303


 Error of  AdaBoostRegressor() :
Mean absolute error : 0.10428942959226736
Mean squared error : 0.017137379542209456
Root Mean Squared Error: 0.13090981453737324


 R2 Score :
0.9450005745499995




As we can see linear regression is giving better result.

# Cross Validation

In [53]:
from sklearn.model_selection import cross_val_score
score = cross_val_score(lin_reg, X_scaler, y, cv = 5)
print('\033[1m'+'Cross Validation Score :'+'\033[0m\n')
print("Score :" ,score)
print("Mean Score :",score.mean())
print("Std deviation :",score.std())

Cross Validation Score :

Score : [0.97465526 0.98274082 0.97968604 0.97846211 0.97135915]
Mean Score : 0.9773806762511942
Std deviation : 0.003972455670334777
